core/unit_test/view/TestReferenceCountedDataHandle.hpp

```diff
@@ -0,0 +1,131 @@
+//@HEADER
+// ************************************************************************
+//
+//                        Kokkos v. 4.0
+//       Copyright (2022) National Technology & Engineering
+//               Solutions of Sandia, LLC (NTESS).
+//
+// Under the terms of Contract DE-NA0003525 with NTESS,
+// the U.S. Government retains certain rights in this software.
+//
+// Part of Kokkos, under the Apache License v2.0 with LLVM Exceptions.
+// See https://kokkos.org/LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//@HEADER
+
+#include <Kokkos_Core.hpp>
+
+#include <gtest/gtest.h>
+
+/*
+template <class ElementType, class MemorySpace>
+class ReferenceCountedDataHandle {
+ public:
+  using value_type   = ElementType;
+  using pointer      = value_type*;
+  using reference    = value_type&;
+  using memory_space = MemorySpace;
+  ...
+}
+*/
+
+namespace {
+using element_t = float;
+using mem_t     = typename TEST_EXECSPACE::memory_space;
+using data_handle_t =
+    Kokkos::Impl::ReferenceCountedDataHandle<element_t, mem_t>;
+using const_data_handle_t =
+    Kokkos::Impl::ReferenceCountedDataHandle<const element_t, mem_t>;
+;
+}  // namespace
+
+TEST(TEST_CATEGORY, RefCountedDataHandle_Typedefs) {
+  Kokkos::parallel_for(
+      Kokkos::RangePolicy<TEST_EXECSPACE>(0, 1), KOKKOS_LAMBDA(int) {
+        static_assert(
+            std::is_same_v<typename data_handle_t::value_type, element_t>);
+        static_assert(
+            std::is_same_v<typename data_handle_t::pointer, element_t*>);
+        static_assert(
+            std::is_same_v<typename data_handle_t::reference, element_t&>);
+        static_assert(
+            std::is_same_v<typename data_handle_t::memory_space, mem_t>);
+      });
+}
+
+TEST(TEST_CATEGORY, RefCountedDataHandle) {
+  auto shared_alloc =
```

Comment:
For such tests, I would have designed it with a test suite setup that makes the shared allocation record, and that would allow me to split the tests in smaller bits that are more readable and on which it's easier to reason about.

Reply:
I thought about it, but it's still a lot of duplicated code. Essentially what it comes down to is that the only way state of RefCountedDataHandle changes is through constructors. So it all comes down to: did the constructor do what we expected it to do. Furthermore for almost anything I need to start with an allocation, so might as well use this one ...

Reply:
You could probably initialize your allocation once (drafter pseudo-code):
```c++
class TestSuite
{
void setup() { if (!data) make share alloc;}
static data = nullptr;
}
```

Reply:
Actually that doesn't work. When you hand the thing to the data handle you pass over ownership. So it will delete the allocation when it goes out of scope.

Reply:
(or when the last one with a ref count goes out of scope)

Reply:
Because you are initializing `Kokkos` before `GTest` in `core/unit_test/UnitTestMainInit.cpp`:
```c++
class TestSuite : public ::testing::Test
{
public:
    static void SetUpTestSuite() {
        if( ! shared_alloc) shared_alloc = Kokkos::Impl::make_shared_allocation_record<...>(...);
    }
    static void TearDownTestSuite() { /* anything to do here ? */}
protected:
    Kokkos::Impl::SharedAllocationRecord<MemorySpace, functor_type>* shared_alloc = nullptr;
};
```

Reply:
And if you should do anything at all, at least do (enhances test readability):
```c++
auto* shared_alloc = ...
```

Reply:
Yeah it returns a pointer to a SharedAllocationRecord, but as soon as I pass that guy to the data_handle constructor the ownership of the allocation the record points to transfers. I.e. I can't pass the same pointer I get back to more than one data_handle thing. And that would break the suggestion above wouldn't it?